### DISCUSSION TITLE
Transition plan for eslint-plugin-json users

### DISCUSSION THREAD

**Discussion by azeemba** [eslint/json#24] [General]
I am the owner of eslint-plugin-json which was always a hacky solution so I am happy to see a proper eslint-maintained solution for JSON.

I am hoping we can discuss and think through a transition plan for users. One option could be for the eslint org to take over the eslint-plugin-json name and maintain it as an alias but I don't know how that would work or what problems it could have. 

Also would be good to know what timelines are being considered to mark this plugin as v1 stable. 

Excited to figure out a plan here!

**Comment by nzakas** [comment]
Hi @azeemba thanks for stopping by, and thanks for `eslint-plugin-json`. It definitely opened up the possibility of linting JSON in ESLint at a time when we weren't quite sure how to do it.

What would be most helpful to get started is if you could take a look at `@eslint/json` and let us know if there is any functionality that is currently in `eslint-plugin-json` that is not in `@eslint/json`. When we ask people to transition, I'd like to make sure they aren't losing any functionality they were depending on.

I think it would be best for you to keep `eslint-plugin-json`. There isn't much value in moving it into the ESLint org if it will just be deprecated.

With regards to timelines: we are formally announcing support as soon as we are able to launch our new code explorer so people can see the ASTs that `@eslint/json` and `@eslint/markdown` create. I'm hoping that will happen with the next two weeks. Then, we'll announce `@eslint/json` and `@eslint/markdown` as ready to try out and provide feedback. For `@eslint/json`, the path to v1.0.0 probably means ensuring we've got compatibility with both `eslint-plugin-json` and `eslint-plugin-jsonc` so people have a smooth transition. So, there's no set timeline, it will depend on the feedback we receive.

**Reply by MikeMcC399** [reply]
According to https://github.com/eslint/json/blob/main/README.md#rules it contains 2 rules.

https://github.com/azeemba/eslint-plugin-json/blob/master/README.md#individual-rules lists many more rules.

How can `eslint/json` replace `eslint-plugin-json` if it doesn't provide at least the same set of rules? What am I missing here? It doesn't seem to make sense to recommend anybody to transition.

**Reply by nzakas** [reply]
Did you read further down on the README?

https://github.com/eslint/json/blob/main/README.md#what-about-missing-rules-that-are-available-in-eslint-plugin-json-and-eslint-plugin-jsonc

**Reply by MikeMcC399** [reply]
@nzakas Thank you! That was the missing piece.

**Comment by azeemba** [comment]
I will reiterate what @MikeMcC399 highlighted but in a broader context: `eslint-plugin-json` will flag more specific errors that currently `@eslint/json` collapses into a single `Parsing error`.

This means if there are multiple issues with a JSON file, then only the first error get reported by `@eslint/json`. For example, here is a JSON file I tested with (it has a missing comma, undefined value and an invalid escape string):

```
{
  "a": 123
  "b": c,
  "\k": 1
}
```

This is the output from `@eslint/json`:

```
/home/z/local-projects/eslint-json-migration-testing/test.json
  3:8  error  Parsing error: Unexpected character 'c' found

✖ 1 problem (1 error, 0 warnings)
```

This is the output from `eslint-plugin-json`:

```
/home/z/local-projects/eslint-json-migration-testing/test.json
  3:4  error  Expected comma                      json/*
  3:9  error  Value expected                      json/*
  4:4  error  Invalid escape character in string  json/*

✖ 3 problems (3 errors, 0 warnings)
```

This makes `eslint-plugin-json` more valuable while the user is actively editing a JSON file. If folks are depending on this in their IDE set up, it may make the migration harder.

**Reply by MikeMcC399** [reply]
@azeemba 
Thank you very much for the concrete example! Based on that, I would want to stay with `eslint-plugin-json` and not migrate.

**Reply by nzakas** [reply]
Thanks for the feedback. Just to clarify for others who come across this thread: `eslint-plugin-json` is not actually flagging more problems. Both plugins flag the same problems, it's just that `@eslint/json` stops at the first parsing error (which is common for parsers and the way the JavaScript parser works in ESLint).

ESLint itself doesn't actually support recoverable parsing errors at this point, so it would take both updating the parser and the ESLint core to make this happen. It's definitely something worth considering.
